### PR TITLE
feat(api-client): edit nested multipart object properties as individual form rows

### LINKED ISSUE #4834: Scalar misinterprets multipart/form-data with part that contains structured JSON
### What happens?

We have an endpoint that accepts a multipart/form-data body with two parts:

* an unstructured file, named `file`
* a structured JSON document, named `props`

The OpenAPI document describes multipart/form-data as `requestBody` with two properties: `file` of type `string` (and format `binary`), and `props` of type `object`, with its own properties. The minimal schema that reproduces this is [attached here](https://github.com/user-attachments/files/18891209/minimal-schema.json), and also [at the sandbox](https://sandbox.scalar.com/e/2Hwa2).

Scalar 0.3.150, which we currently deploy, generates the following request:

```sh
curl URL \
  --request POST \
  --header 'Content-Type: multipart/form-data' \
  --form 'file=' \
  --form 'props[created_at]=null' \
  --form 'props[description]=' \
  --form 'props[name]='
```

It is incorrect because it splits up `props` into multiple form parts instead of as a single parts encoded in JSON. It seems to use [qs](https://github.com/ljharb/qs)-like syntax, which was not requested by the schema and won't be accepted by the endpoint.

The scalar deployed on https://docs.scalar.com/swagger-editor generates the following:

```sh
curl URL \
  --request POST \
  --header 'Accept: */*' \
  --header 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundarybtmt1870GXquBkll' \
  --data "{"file":"","props[name]":"","props[description]":"","props[created_at]":""}"
```

This is also incorrect, and a step back from the older version because:

* it manually specifies form data boundary instead of leaving encoding up to curl
* it sends the whole form in a single JSON-encoded part instead of in two parts
* the keys of `props` are again encoded like with qs

### What did you expect to happen?

Correct request might look like this:

```sh
curl URL \
  --request POST \
  --header 'Content-Type: multipart/form-data' \
  --form 'file=@filename' \
  --form 'props={"name": "", "created_at": "", "description": ""}'
```

It is generating two form parts, one with the raw contents of a file (but `'file=whatever` would also be fine), and the other with `props` initialized to JSON valid for the schema.

While there are already several issues about multipart/form-data, they don't seem to cover the exact same bug. In case it's relevant, [swagger's editor](https://editor-next.swagger.io/) interprets the document as intended.

### OpenAPI Document

https://sandbox.scalar.com/e/2Hwa2

### PR DESCRIPTION
**Pull request by hanspagel**
Nested object properties in `multipart/form-data` schemas used to collapse into a single JSON-stringified field. Now each leaf gets its own editable row (`props.name`, `props.description`, `props.created_at`), and dotted rows are regrouped into one `props` multipart part on send.

cURL preview also pretty-prints JSON inside `--form` parts, matching what we already do for `--data`.

**Before**
Screenshot 2026-05-11 at 11 26 02

**After**
Screenshot 2026-05-11 at 12 35 08

Fixes #4834

## OpenAPI Document

```yaml
openapi: 3.1.0
info:
  title: Multipart Nested Rows Test
  version: 1.0.0
paths:
  /widgets:
    post:
      summary: Create widget (multipart)
      requestBody:
        required: true
        content:
          multipart/form-data:
            schema:
              type: object
              required: [file, props]
              properties:
                file:
                  type: string
                  format: binary
                  description: File to upload
                props:
                  type: object
                  required: [name, description]
                  properties:
                    name: { type: string }
                    description: { type: string }
                    created_at: { type: string, format: date-time }
      responses:
        '200': { description: OK }

  /widgets-url:
    post:
      summary: Create widget (urlencoded) — same schema as multipart
      requestBody:
        required: true
        content:
          application/x-www-form-urlencoded:
            schema:
              type: object
              required: [token, props]
              properties:
                token: { type: string}
                props:
                  type: object
                  properties:
                    name: { type: string}
                    description: { type: string}
      responses:
        '200': { description: OK}
```

---

> [!NOTE]
> **Medium Risk**
> Moderate risk because it changes how multipart form rows are generated and serialized, including schema-driven regrouping of dotted keys into JSON parts; mistakes could alter request payload shape or field ordering.
> 
> **Overview**
> Enables **nested `multipart/form-data` object properties** to be edited as individual dotted-name rows (e.g. `props.name`) by walking the request-body schema + example and emitting one row per leaf, while preserving per-leaf schema metadata and *cascading* `required` semantics.
> 
> Updates request-body building for multipart so **edited dotted rows are regrouped back into a single top-level JSON text part** (one part per top-level object property), preserving row order and avoiding accidental folding of literal dotted names (e.g. `user.email`) unless the schema declares a nested object. Adds comprehensive tests for nesting depth, null/file handling, undeclared example keys, urlencoded non-expansion, and dotted-name edge cases; also tightens HAR param building by normalizing row-array names to strings.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 352faf2678c76db3caf7d55cddbfdc62f3c1a9ff. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/api-client/src/v2/blocks/operation-code-sample/helpers/operation-to-har/process-body.ts [hunk 1]
```diff
@@ -31,8 +26,8 @@ const objectToFormParams = (
   const params: Param[] = []
 
   /** Ensure we do not include disabled items */
-  const entries = Array.isArray(obj)
-    ? obj.filter((item) => !item.isDisabled).map((item) => [item.name, item.value])
+  const entries: [string, unknown][] = Array.isArray(obj)
+    ? obj.filter((item) => !item.isDisabled).map((item) => [String(item.name), item.value] as [string, unknown])
```
[the remark this answers is not in the document]
Comment (hanspagel) [reply packages/api-client/src/v2/blocks/operation-code-sample/helpers/operation-to-har/process-body.ts]: covered in #9145

File: packages/api-client/src/v2/blocks/request-block/helpers/get-form-body-rows.ts [hunk 2]
```diff
@@ -16,6 +17,39 @@ const stringifyValue = (value: unknown) => {
...
+  const requiredSet = new Set(schema.required ?? [])
+  const leaves: LeafRow[] = []
+  for (const [key, rawChildSchema] of objectEntries(schema.properties)) {
+    const childSchema = resolve.schema(rawChildSchema)
+    const path = [...parentPath, String(key)]
+    const isRequired = parentRequired && requiredSet.has(String(key))
+    if (childSchema && isObjectSchema(childSchema) && childSchema.properties) {
+      leaves.push(...collectLeafProperties(childSchema, path, isRequired))
+    } else {
+      leaves.push({ path, schema: childSchema, isRequired })
```
Comment (amritk) [inline packages/api-client/src/v2/blocks/request-block/helpers/get-form-body-rows.ts:57]: would be nice to space this out with some comments

File: packages/api-client/src/v2/blocks/operation-code-sample/helpers/operation-to-har/process-body.ts (hunk 1 above)
Comment (amritk) [inline packages/api-client/src/v2/blocks/operation-code-sample/helpers/operation-to-har/process-body.ts]:
```suggestion
    ? obj.filter((item) => !item.isDisabled).map((item) => [String(item.name), item.value])
```

no need

**Review by amritk — CHANGES_REQUESTED** [review]
If there's one thing I hate, its nested loops! It should be ok depending on the example size but lets flatten it out if we can, cursor should be able to do it. 
```
    const dottedRows = exampleValue.filter(({ name, value }) => isDottedNestedRow(name, value))
```

**Review by amritk — APPROVED** [review]
cool I think its good

**Comment by marlon-claus** [comment]
This doesn't seem to work correctly anymore. It's split into multiple rows correctly but on send the properties don't seem to be merged into a json object on our end. This causes an error because our api needs one multipart row that contains a json string and not multiple ones.
Could this be a problem on our end? :D
